design_documents/cairo_deployment_scripts.md

```diff
@@ -0,0 +1,336 @@
+We will use contracts dispatchers to be able to call/invoke its functions directly 
+(e.g. if contract named `mycontract` has a function `myfunction`, we should be able to just do `mycontract.myfunction();`).
+Later on we could also support our subcommands (`invoke`, `call`) to call/invoke contracts without dispatchers.
+
+### running the script
+The script would essentially run an entrypoint function - `main`. Inside our script subcommand, we will have to compile 
+the cairo script file using `scarb build`, and then run it using `scarb cairo-run`. 
+The main function would be required in the deployment script, and we should return an error if it is not found.
+
+In order to build the script, there'll need to be a `Scarb.toml` file present in the working directory. We should either:
+- look for a Scarb.toml in the same directory script is (this is scarb's default behaviour - we just must make sure scarb
+command is executed from this directory - Command::new(...).current_dir(...) should do); if does not exist we could just create it
+- create a hidden Scarb.toml with a custom name (eg `.script_name.toml`) and override path to scarb toml using scarb flag;
+if a file does not exist let's just create it
+
+Option 1 is probably more in line with scarb ethos, but the option 2 would probably be more practical for us as it would allow
+for multiple Scarb.toml files in one directory, where multiple scripts could reside + Scarb.toml file present in cwd would not
+confuse sncast subcommands (eg declare also looks for Scarb.toml in cwd, which could lead to errors/user confusion).
+
+The deployment script could then be run like so:
+```bash
+$ sncast script /path/to/myscript.s.cairo
+```
+
+Log info should be outputted to stdout and optionally written to some log file. 
+We could have a flag to suppress outputting to stdout or to output with different formats (int, hex).
+
+By default, the scripts could run in some kind of "dry run mode", meaning all the transactions would be simulated with 
+specific rpc calls (like `starknet_simulateTransactions`). An interesting idea would be to research an approach similar 
+to cheatnet - fork a runner and test the scripts against it (to be assessed if it is feasible). We could also provide 
+a way to show a trace of said transactions if required. In later iterations, we could also provide a helper function
+that sets up the devnet to be used for dry running for users.
+
+The behaviour would be changed with `--broadcast` flag, that would actually execute needed transactions.
+
+### error handling
+If the transaction fails, we should put all the relevant info about it to the state file, output log information and stop
+script execution.
+If the script fails (for any reason that is not connected to transactions - eg rpc node down), we should output relevant
+log information.
+
+In case of failed transaction, in the "output" field in state file we could include an error message. After fixing the issue
```

Comment:
We should probably ask the user if he wants to re-run from the end of the last txn (i.e. `--continue`)  or run a new flow (y/n confirmation?).

Reply:
Wouldn't be better to let users handle errors themselves in cairo with Result?

Reply:
> We should probably ask the user if he wants to re-run from the end of the last txn (i.e. --continue) or run a new flow (y/n confirmation?).

re-running from last tx would be a default action, as the script aims to be idempotent. There is a flag `--replay` that runs a new flow. Is that ok?

> Wouldn't be better to let users handle errors themselves in cairo with Result?

Hmm, makes sense, but I would also like to keep current behaviour as a possibility. I can imagine a scenario, where someone wants the script to just fail if a transaction fails, and be able to act on it later. would that be ok? + included your suggestion in newest commit.